internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+	// Generate
+	keypair, err := m.genECP256Keypair(notBefore, notAfter)
+	if err != nil {
+		return nil, err
+	}
+	// Persist
+	err = m.persistTLSKeyPair(keypair)
+	if err != nil {
+		return nil, err
+	}
+	m.tlsKeyPairs = append(m.tlsKeyPairs, keypair)
+	// Sort by valid starting date, end date and fingerprint.
+	slices.SortFunc(m.tlsKeyPairs, func(a, b *TLSKeyPair) int {
+		return a.Compare(b)
+	})
+	return keypair, nil
+}
+
+// genECP256Keypair generates a new keypair based on ECDSA NIST P-256
+// private key and X509 public certificate. The certificate is signed with
+// the same private key (self-signed), which means that the public key
+// in the certificate can be used to verify the signature.
+func (m *TLSManager) genECP256Keypair(notBefore time.Time, notAfter time.Time) (*TLSKeyPair, error) {
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	// Random serial for now.
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, err
+	}
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		NotBefore:    notBefore,
+		NotAfter:     notAfter,
+		Subject: pkix.Name{
+			Organization: []string{"Canonical Ltd."},
+		},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+	// DER encoded bytes
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &privateKey.PublicKey, privateKey)
+	if err != nil {
+		return nil, err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	certificate, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return nil, err
+	}
+	keypair := &TLSKeyPair{privateKey: privateKey, Certificate: certificate}
+	keypair.addFingerprint()
+	return keypair, nil
+}
+
+// persistTLSKeyPair persists the TLS keypair in a single PEM file. The file name
+// is the certificate fingerprint string, with a .pem suffix. See the
+// addFingerprint method for information on how the fingerprint is generated.
+func (m *TLSManager) persistTLSKeyPair(keypair *TLSKeyPair) (err error) {
+	// If the TLS directory does not yet exist, create it.
+	_, err = os.Stat(m.tlsDir)
+	if os.IsNotExist(err) {
+		// Create the leaf directory node with 0700 permissions.
+		err = os.Mkdir(m.tlsDir, 0700)
+	}
+	if err != nil {
+		return fmt.Errorf("cannot create directory leaf of path %v: %w", m.tlsDir, err)
+	}
+
+	pemPath := filepath.Join(m.tlsDir, fmt.Sprintf("%v.pem", keypair.Fingerprint()))
+	pemFile, err := os.OpenFile(pemPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
```

Comment:
Calling both `Sync` and `Close` shouldn't be necessary. `Close` will implicitly synchronize.

Reply:
Close does not imply sync last time I checked. It allows some users how do not care about lazy cache draining to leave it to the OS and continue with more important stuff.

Reply:
Resolving.